lib/util/location.ex

```diff
@@ -97,6 +97,89 @@ defmodule Util.Location do
     distance(as_location!(lhs), as_location!(rhs))
   end
 
+  @doc """
+  Returns the distance between the point and the segment defined by
+  segment_start and segment_end, as well as the point on the segment
+  that provides that minimum distance.
+
+  ## Examples
+      iex> Util.Location.distance_from_segment(
+      ...>   Util.Location.new(42.00004, -71.00014),
+      ...>   {Util.Location.new(42, -71.0002), Util.Location.new(42.0001, -71.0001)}
+      ...> )
+      %{
+        closest_point: %Util.Location{latitude: 42.00004711559566, longitude: -71.00015288440434},
+        distance: 1.3264961988869075
+      }
+  """
+  @spec distance_from_segment(
+          point :: __MODULE__.From.t(),
+          {segment_start :: __MODULE__.From.t(), segment_end :: __MODULE__.From.t()}
+        ) :: %{closest_point: __MODULE__.t(), distance: number()}
+
+  def distance_from_segment(
+        %__MODULE__{latitude: latitude, longitude: longitude} = point,
+        {%{latitude: start_latitude, longitude: start_longitude},
+         %{latitude: end_latitude, longitude: end_longitude}}
+      ) do
+    # Scale factors used to transform between lat/long and a coordinate
+    # system based on distance
+    latitude_scale_factor =
+      1000 *
+        distance(
+          new(start_latitude, start_longitude),
+          new(start_latitude + 0.001, start_longitude)
+        )
+
+    longitude_scale_factor =
+      1000 *
+        distance(
+          new(start_latitude, start_longitude),
+          new(start_latitude, start_longitude + 0.001)
+        )
+
+    # Use the above scale factors to transform both the original
+    # segment and the point in question to (x, y) coordinates, where
+    # the origin is the segment start point, x is meters east from
+    # there, and y is meters north
+    segment_end_x = longitude_scale_factor * (end_longitude - start_longitude)
+    segment_end_y = latitude_scale_factor * (end_latitude - start_latitude)
+
+    x = longitude_scale_factor * (longitude - start_longitude)
+    y = latitude_scale_factor * (latitude - start_latitude)
+
+    squared_segment_length = segment_end_x ** 2 + segment_end_y ** 2
+
+    # Transform the coordinates of the non-segment point to (u, v),
+    # where segment_start is at (0, 0) and segment_end is at (1, 0).
+    # We don't bother computing the v coordinate because we don't need
+    # it.
+    u = (x * segment_end_x + y * segment_end_y) / squared_segment_length
```

Comment:
I decided to structure it this way because I found it easier to understand as a series of two transformations (and then reversing those two transformations at the end).

I could also consolidate it into a single formula if that would be easier for others to understand.

Also, @firestack mentioned that this would use a dot product, and I was like "noooo", but lo and behold - this formula is indeed a dot product!

Reply:
I haven't had the chance to really read over this and understand it, though I've been trying. Given that this ended up being the dot product, could the dot product be extracted into a function to hide the operation? I personally think it'd read easier, but I'm not sure what it'll look like after that and could easily be wrong on this.